src/astarte.rs

```diff
@@ -0,0 +1,441 @@
+use astarte_device_sdk::transport::mqtt::{Credential, Mqtt, MqttConfig};
+use astarte_device_sdk::{Client, DeviceClient, DeviceConnection};
+use clap::ValueEnum;
+use color_eyre::eyre;
+use color_eyre::eyre::{eyre, ContextCompat, WrapErr};
+use serde::Deserialize;
+use std::path::{Path, PathBuf};
+use std::time::SystemTime;
+use std::{env, io};
+use tracing::{debug, error, info};
+
+/// Stream Rust test node identifier
+const STREAM_RUST_TEST_NODE_UUID: uuid::Uuid = uuid::uuid!("d72a6187-7cf1-44cc-87e8-e991936166dc");
+
+const DEVICE_DATASTREAM: &str =
+    include_str!("../interfaces/org.astarte-platform.genericsensors.Values.json");
+
+/// Specify which Astarte library use to connect to Astarte
+#[derive(
+    Debug, Clone, Copy, Default, Eq, PartialEq, Ord, PartialOrd, clap::ValueEnum, Deserialize,
+)]
+pub enum AstarteConnection {
+    /// Connect through MQTT to Astarte
+    #[default]
+    #[serde(rename = "mqtt")]
+    Mqtt,
+    /// Connect through gRPC to an Astarte Message Hub instance already connected to Astarte
+    #[serde(rename = "grpc")]
+    Grpc,
+}
+
+/// Astarte connection config toml
+///
+/// struct used to deserialize the TOML file
+#[derive(Debug, Default, Deserialize)]
+struct ConfigToml {
+    astarte: ConnectionConfigBuilder,
+}
+
+/// Astarte connection config Builder
+#[derive(Debug, Default, Deserialize)]
+pub struct ConnectionConfigBuilder {
+    /// Astarte connection
+    #[serde(rename = "connection")]
+    astarte_connection: Option<AstarteConnection>,
+    /// Astarte store directory
+    store_directory: Option<PathBuf>,
+    /// Astarte Device SDK config options
+    #[serde(rename = "mqtt", default)]
+    mqtt_config: MqttConfigBuilder,
+    /// Astarte Message Hub config options
+    #[serde(rename = "grpc", default)]
+    grpc_config: GrpcConfigBuilder,
+}
+
+impl ConnectionConfigBuilder {
+    /// Builder constructor
+    ///
+    /// Specify if the builder should use the Astarte Device SDK or the Astarte Message Hub
+    pub fn with_connection(astarte_connection: Option<AstarteConnection>) -> Self {
+        Self {
+            astarte_connection,
+            store_directory: None,
+            mqtt_config: MqttConfigBuilder::default(),
+            grpc_config: GrpcConfigBuilder::default(),
+        }
+    }
+
+    /// Init astarte config from env var if they have been set
+    pub fn from_env(&mut self) {
+        // doesn't change it if it's been set from CLI
+        if self.astarte_connection.is_none() {
+            self.astarte_connection = env::var("ASTARTE_CONNECTION")
+                .ok()
+                .map(|s| AstarteConnection::from_str(&s, true))
+                .transpose()
+                .ok()
+                .unwrap_or_default();
+        }
+
+        self.store_directory = env::var("ASTARTE_STORE_DIRECTORY").ok().map(PathBuf::from);
+
+        // update the mqtt config info
+        let device_id = env::var("ASTARTE_DEVICE_ID").ok();
```

Comment:
If the `connection` selected is `grpc` you can skip to read the parameter for `mqtt` connection.

Reply:
At the moment we can set  the type of connection with the CLI, through environment variables or by using a `config.toml` file. The ideal order from which we take the configuration parameters is CLI -> ENV VARS -> toml. However, in the actual implementation, some information can be taken form the CLI (e.g., the type of connection), others from ENV and others from toml. In the end, it will be the `build` method (of the `ConnectionConfigBuilder`) which will verify that all the necessary information to establish a connection to astarte has been properly set. However, since the astarte connection type (mqtt or grpc) can be set in the last step (that is, when reading the config.toml) but other information could have been set through ENV VARS, we cannot avoid reading all the parameters, both for mqtt and grpc.